### PR TITLE
Update mimalloc to 2.1.7

### PR DESCRIPTION
**Pull request by kleisauke**

### CODE REVIEW

**Comment by sbc100** [comment]
Oh, I guess we should make `__builtin_thread_pointer` work.  I've not seen that before.

**Comment by sbc100** [comment]
Could you open an upsteam llvm issue regarding `fatal error: error in backend: Cannot select: intrinsic %llvm.thread.pointer
`?

File: system/lib/mimalloc/src/prim/emscripten/prim.c [hunk 1]
```diff
@@ -72,30 +72,14 @@ extern void* emmalloc_memalign(size_t, size_t);
...
-  } else if (try_alignment > MAX_EMMALLOC_ALIGN) {
-    try_alignment = MAX_EMMALLOC_ALIGN;
-  }
-  void* p = emmalloc_memalign(try_alignment, size);
-  *addr = p;
-  if (p == 0) {
-    return ENOMEM;
-  }
-  return 0;
+  *addr = emmalloc_memalign(try_alignment, size);
```
Comment (kripken) [inline system/lib/mimalloc/src/prim/emscripten/prim.c:81]: Did the emmalloc alignment limitations get fixed?

**Comment by kleisauke** [comment]
Revised this PR to update mimalloc to 2.1.7 instead. This is ready for review now.

> Could you open an upsteam llvm issue regarding `fatal error: error in backend: Cannot select: intrinsic %llvm.thread.pointer `?

I just opened issue https://github.com/llvm/llvm-project/issues/117433 for this. Commit 6906868995286f158d5d34b224b73fbe4edfb6fb defines `MI_LIBC_MUSL` to workaround this.

File: system/lib/mimalloc/src/prim/emscripten/prim.c (hunk 1 above)
Comment (kleisauke) [reply system/lib/mimalloc/src/prim/emscripten/prim.c:86]: This was fixed via PR #21905 in the meantime. I also removed `MIN_EMMALLOC_ALIGN` initially, but let's leave that for now.

File: tools/system_libs.py [hunk 2]
```diff
@@ -1800,6 +1803,11 @@ class libmimalloc(MTLibrary):
     '-DMI_MALLOC_OVERRIDE',
     # TODO: add build modes that include debug checks 1,2,3
     '-DMI_DEBUG=0',
+    # disable `assert()` in emmalloc
+    '-DNDEBUG',
+    # avoid use of `__builtin_thread_pointer()`
+    # FIXME: https://github.com/llvm/llvm-project/issues/117433
```
Comment (sbc100) [inline tools/system_libs.py:1809]: I'm not sure we want a FIXME here since we are using musl as our libc so this seems correct in any case.

Comment (kleisauke) [reply tools/system_libs.py]: a090ba89182812faed943b864fff8b7046a8e6fa

File: tools/system_libs.py [hunk 3]
```diff
@@ -1800,6 +1803,10 @@ class libmimalloc(MTLibrary):
     '-DMI_MALLOC_OVERRIDE',
     # TODO: add build modes that include debug checks 1,2,3
     '-DMI_DEBUG=0',
+    # disable `assert()` in emmalloc
```
Comment (kripken) [inline tools/system_libs.py:1806]:
```suggestion
    # disable `assert()` in mimalloc
```
But do we not want this in debug builds?

Comment (sbc100) [reply tools/system_libs.py:1806]: If we do, I think that we can/should make that a separate change.

Comment (kleisauke) [reply tools/system_libs.py:1806]: I think the comment is correct in this case, mimalloc uses `mi_assert`, which is already disabled.
https://github.com/microsoft/mimalloc/blob/v2.1.7/include/mimalloc/types.h#L563-L569

However, we also want to disable the assertions in the underlying emmalloc allocator, which we build as part of mimalloc:
https://github.com/emscripten-core/emscripten/blob/a090ba89182812faed943b864fff8b7046a8e6fa/tools/system_libs.py#L1818-L1819
https://github.com/emscripten-core/emscripten/blob/a090ba89182812faed943b864fff8b7046a8e6fa/tools/system_libs.py#L1828
https://github.com/emscripten-core/emscripten/blob/a090ba89182812faed943b864fff8b7046a8e6fa/system/lib/emmalloc.c#L33-L35
https://github.com/emscripten-core/emscripten/blob/a090ba89182812faed943b864fff8b7046a8e6fa/system/lib/libc/musl/include/assert.h#L5-L9

Comment (kripken) [reply tools/system_libs.py:1806]: Oh, it's for the underlying allocator. In that case it sounds ok. But can it be a separate change, or was it necessary for some reason to do in this PR?

Comment (kleisauke) [reply tools/system_libs.py:1806]: There's no specific reason to address it in this PR; it was discovered by accident while working on this. Split into PR #23016.

**Review by kripken — APPROVED** [review]

**Comment by kripken** [comment]
Thanks @kleisauke !

**Pull request merged**